### ISSUE TITLE
async-node-http chunk loading

### ISSUE THREAD

**Issue by 7rulnik** [webpack/webpack#12084]
## Feature request

**What is the expected behavior?**

It should be possible to have different servers and connect them via module federation over http, not fs.

**What is motivation or use case for adding/changing the behavior?**

I was playing some time with module federation and realized that all examples for server-side use the same file system. I think that very naive representation and usually, we use something like docker containers at least. It would be great to have a way to load chunks from a completely different server via http.

**How should this be implemented in your opinion?**

Basically, `async-node` target contains everything that we need. We just need change `require('fs').readFile` logic to use HTTP agent.

Also, there is a question of how extendable HTTP agent should be? Because some people may want to use authentication for example.

And it's not clear who should be responsible for launching HTTP server for serving chunks? Should it be done in userland or seted up by webpack runtime?

**Are you willing to work on this yourself?**
yes

------------

I already have a working example. It still WIP but it works. To be honest I'm not sure that it should be in webpack core because it feels kinda hacky and as I understand ESM should allow the same thing in a more robust way (and I saw that it will be introduced in v5).

**Comment by alexander-akait** [comment]
I think it is out of scope webpack, you can implement own plugin for this, we provide rich API

**Comment by sokra** [comment]
I recommend to copy the code from remotes to the local filesystem before starting the server. e. g. via a package manager like yarn or npm. That's a lot safer than to notice at runtime that a file from remote server can't be fetched due to whatever network thing.

**Comment by 7rulnik** [comment]
@sokra I agree that it's a more risky way.

What if I want to make a zero downgrade update? I'm thinking about calling a package manager from a node process and using hot reload in production mode? What do you think?

**Comment by alexander-akait** [comment]
> using hot reload in production mode

Bad idea

**Comment by 7rulnik** [comment]
@alexander-akait excuse me, but it's not a really useful comment. Of course, I saw `never use hmr in prod` but what real limitations and problems behind that?

**Comment by sokra** [comment]
> Of course, I saw `never use hmr in prod` but what real limitations and problems behind that?

HMR is also possible in production. It's risky but the feature was always intended to also work in production.

> What if I want to make a zero downgrade update? I'm thinking about calling a package manager from a node process and using hot reload in production mode? What do you think?

While technically possible, it's pretty risky. You need to really test HMR of the code. Best make sure to only `apply` the update when your server is in idle anyway.

Usually one would spawn a new server with the new code and swap servers on load balancer level, but this really depends on you application. If you need to apply server updates while long running connections stay open that might not work for you.

When using a package manager you might run into a problem: the package manager would remove the old files, but they might be still needed by the running server. So you need to make sure to not remove old files. And all (on demand loaded) files need to have unique names. So use best use contenthashing for all files.

Also note that module federation doesn't support HMR between builds yet. So currently you need to handle HMR within builds. So currently I can't recommend to use HMR with module federation except for some cases. You need a fallback solution when HMR won't work.

**Comment by alexander-akait** [comment]
bump

**Comment by 7rulnik** [comment]
@telenko published his own plugin and seems that we both agree that it shouldn't be in core so for now I think it should be closed


https://mangolik931.medium.com/how-i-implemented-module-federation-in-nodejs-and-did-something-wrong-724642c26da5
https://github.com/telenko/node-mf
https://github.com/telenko/node-mf-example
